loops/csharp/code.cs

```diff
@@ -0,0 +1,16 @@
+﻿var u = int.Parse(args[0]);
+var r = Random.Shared.Next(10_000);
+
+//Use span instead of array
+Span<int> a = new int[10_000];
+
+//using a.Length instead of 10_000 constant should give the compiler better optimization hints
+for (var i = 0; i < a.Length; i++)
```

Comment:
I forgot to mention but using here a.Length over 10_000 constant has no impact since .NET can see that both are identical. It stopped needing most of such hacks over the years :) If you just have an array using a span is also not necessary since arrays is the primary optimization target - array and span optimizations are usually on par with each other not counting some minor differences in codegen.

Reply:
Shouldn't the array stack allocated? Seems like it's faster as mentioned here #21.

Reply:
No. There would be no difference - it was discussed in a separate PR too: https://github.com/bddicken/languages/pull/25#discussion_r1859485554

In .NET, `Span<T>` and `ReadOnlySpan<T>` are slice-types much like in Rust. What this means is they do not *own* the memory, they only observe it.

Internally, a span holds `ref T _reference` and `int _length`. That said `ref T` can originate from any source of memory which makes .NET a very unique GC-based platform.

Like other types with contiguous memory, `T[]` has an implicit cast to a `Span<T>`. This means creating a span that holds a byref pointer to the first element of the array and its length. The memory that the span observes still has to live somewhere.

Here, `Span<int> a = new int[10_000];` is a variable declaration and assignment statement. The right-hand side expression in this statement is evaluated first, which performs a heap allocation of an `int[]` array, which is then implicitly converted to a span.

Indeed, there are specially recognized patterns where sometimes array literals immediately consumed by `ReadOnlySpan<T>` which are converted to special readonly data without allocation, but it does not apply here.

To summarize, a span itself can only be placed on stack due to the way GC tracking of `ref T` byref pointers works, but the memory that said span observes can be located anywhere.

Reply:
Thank you so much for the explanation—it's incredibly informative and has helped me resolve some struggles I was having with understanding Span<T> and ReadOnlySpan<T>. :)

I did know that both types merely observe the underlying memory, pointing to the first element of the array. And like when you stackalloc something, the memory is cleared from the stack once the method returns.

However, I hadn’t considered what you pointed out here:

> Like other types with contiguous memory, T[] has an implicit cast to a Span<T>. This means creating a span that holds a byref pointer to the first element of the array and its length. The memory that the span observes still has to live somewhere.

If I understand correctly, when an array is allocated, the GC ensures the array remains alive even after the method’s stack frame is popped. Is that accurate?

This also clarifies why the following doesn’t work:
```csharp
static Span<int> GetSpan()
{
    Span<int> span = stackalloc int[256];
    return span;
}
```
While this does:
```csharp
static Span<int> GetSpan()
{
     Span<int> span = new int[256];
     return span;
}
```

> Indeed, there are specially recognized patterns where sometimes array literals immediately consumed by ReadOnlySpan<T> are then converted to special readonly data without allocation, but it does not apply here.

Yes, your right for the `ReadOnlySpan<T>`, As I’ve heard from Stephen Toub, in those cases, the data gets inlined into the data section of the assembly, which makes sense for readonly literals.